### ISSUE TITLE
[technical discovery] 686c | Stop sending mismatched SSNs and file number to BGS

### ISSUE THREAD

**Issue by data-doge** [department-of-veterans-affairs/va.gov-team#56995]
## Value Statement
**_As a_** software developer
**_I want to_** stop sending mismatched SSNs and file numbers to BGS
**_So that_** they don't cause claims to fail due to validation errors.

---
## Background Context

- In https://github.com/department-of-veterans-affairs/va.gov-team/issues/56333, I discovered that we sometimes send a 9-digit ssn to BGS that does not match the veteran's file number. BGS requires the two to match. I took a quick look at one of these instances, and saw that the ssn and file number differed by only one digit, suggesting an input error.
- It appears that we've always had these, and interestingly, we occasionally get huge spikes of these errors.
 ![image](https://user-images.githubusercontent.com/7520103/232609920-48b9a25f-9277-45df-bc01-4c8e3fa81cfd.png)
- See also [Sentry Link](http://sentry.vfs.va.gov/organizations/vsp/discover/results/?environment=production&field=message&name=Top+Errors&project=3&query=level%3Aerror+message%3A%22%2Araise_backend_exception%2ASubmitForm686cJob%22&sort=-message&statsPeriod=90d&widths=-1).
- Looking at the past 90 days, it looks like we get around 9-10 of these every day.

---
## Acceptance criteria
- [x] Stop file number/SSN mismatches where possible 
  - See https://github.com/department-of-veterans-affairs/vets-api/pull/12530
- [ ] Report mismatches that we cant fix to MPI and BGS so that they can follow up
  - Made new ticket for this https://github.com/department-of-veterans-affairs/va.gov-team/issues/57798

## Tasks
- [x] Find where SSNs come from
- [x] Determine if other instances of this error are identical or different
- [x] Confirm with Brandi if there are VBA processes designed to capture these cases (she is investigating)

**Checklist items below can be migrated to https://github.com/department-of-veterans-affairs/va.gov-team/issues/57798** 

## Definition of Ready 
- [ ] Implementation plan is approved by BGS and VBA
- [ ] Dependencies assumption and blockers are listed

## Definition of Done
- [ ] VBA and BGS give sign-off on implementation success
- [ ] Reviewed and approved by product

**Comment by data-doge** [comment]
It turns out the error here is a composite of many other errors. Sentry's grouping of errors, along with the kind of error message we get back from BGS, obscures that. Here's the breakdown of errors for the past 7 days:
```
File # SSN mismatch: 15
File number way too long: 1
File number contains dashes: 5
File number 1 digit too long: 3
Timeout: 6
City too long: 5
Flash error: 4
Invalid data for MLTY_POSTAL_TY: 3
Address Line 1 has invalid chars: 2
ORA-20001: Invalid data for MARAGE_TRMNTN_TYPE_CD = Annulment: 1
Country too long: 1
ORA-02291: integrity constraint (CORPPROD.FK2_VNP_PTCPNT_ADDRS) violated - parent key not found: 1
Dupe claim: 1
City is null: 1
Address Line 2 has invalid chars: 1
The length of the EXTERNAL_UID or EXTERNAL_KEY exceeds the maximum 39 characters allowed: 1
```
Of that breakdown, file # / ssn mismatches are still the most prevalent, and the biggest source of null claims -- a little more than 2 per day. As I look through the code to figure out how to resolve file # / ssn mismatches, I'll also see whether there are any quick fixes for the other errors listed above.

**Comment by data-doge** [comment]
I have a slack thread open in platform support where I've asked whether VA.gov has any insight into the mismatches: https://dsva.slack.com/archives/CBU0KDSB1/p1681247416611139. No answers yet.

I've also reached out to Linda and Baha from BGS, who are unsure why this is happening. They recommended I reach out to Danny Reed from MPI. I reached out to him today and am waiting to hear back.

**Comment by data-doge** [comment]
Had a great conversation with Danny Reed (MPI) today. Some notes:

I sent Danny six instances of file # / ssn mismatches, along with the ICNs associated with the user session. Instance 1, 3, 4, 6 were "off by one" mismatches, and instance 2 and 5 were cases where ssn and file # were substantially different.

Danny was able to determine that instances 1 and 6 were cases where the SSNs had been input into MPI incorrectly, with a typo. He was able to determine that the SSN was incorrect (rather than the file number) by checking the two against SSA (_The_ social security administration) on whatever dashboard he was using. In both of these cases, the previous, incorrect SSN was "not verified." He was also able to correct these SSNs to match the file number.

Instances 2, 3, 4, and 5 were all cases where neither the file number nor ssn I provided were associated with a veteran matching the veteran with the provided ICN. In one case (I believe instance 5), Person A and Person B appeared to be related. They had the same last name and at one point lived at the same address. Danny said he remembered an issue with ID.me where ID.me would have a different SSN than MPI, and that this SSN would belong to a relative of the veteran. I believe Danny noted that in instance 5, the user logged in via DSLOGON.

(more notes forthcoming)

**Comment by data-doge** [comment]
Other questions for BGS:
- Why does `find_person_by_ptcpnt_id` endpoint sometimes return a person with a file # containing dashes (e.g. XXX-XX-XXXX)?
- Why does `find_person_by_ptcpnt_id` endpoint sometimes return a person with a file # with 10 digits?
- Why does `find_person_by_ptcpnt_id` endpoint sometimes return a person with a file # with 13 digits and a dash (e.g. (XXXXXXXX-XXXXX)?

**Comment by data-doge** [comment]
Here's an update and summary of the technical discovery on this ticket, that I posted earlier in Slack:

**1. There are issues with the SSNs we get from MPI and VA identity providers that we don't have much control over.**

Out of the sample of mismatches I have been investigating, about 60% are cases where the SSN we get from MPI or another VA identity provider (e.g. ID.me, DSLogon) are incorrect. In some of these cases, the SSN in MPI just had a little typo. In other cases, the SSN from a VA identity provider belonged to a completely different person (in one of those cases, it appeared that the SSN belonged to a relative of the veteran).

Danny at MPI and Trevor from the Identity Team are both aware of this problem with bad SSN (and other) data. Trevor and the Identity Team have handled this issue by 1) writing code to detect veterans with screwed up data when they log in, 2) compiling reports of these veterans, and 3) sending these reports to MPI to resolve. Trevor explained that MPI has a team of people who are dedicated to manually reconciling screwed up veteran data.

This "bad data" problem is a big one and is largely beyond my control and beyond the scope of Form 686c.  Though, we could try detecting, compiling, and reporting screwed up veteran accounts to MPI to resolve, like the Identity Team does. We could also ask the Identity Team to do that for us. Alternatively, we could "fail sooner."  For instance, we could prevent the veteran from filling out Form 686c if their data is screwed up. We already do something like this for COE and other forms. For example, for COE, if a veteran does not have an EDIPI, we block access to the COE form and instead show an alert that tells the veteran to contact VA's helpdesk.

**2. There are issues with the File #s we get from BGS that we don't have control over.**

Out of the sample of SSN / File # mismatches I have been investigating, about 40% are cases where the File # that we get from BGS (via the find_person_by_ptcpnt_id endpoint) are "bad."  The "bad" file numbers fall into three categories: 1) seemingly legitimate file #, but contains dashes (e.g. XXX-XX-XXXX), 2) 10-digit file number (XXXXXXXXXX), and 3) a date and a number (e.g. YYYYMMDD-XXXXX).

According to Cory at BGS, the file number we get back should be a nine or eight-digit number with no dashes.  Cory doesn't know where this bad data comes from. We're both planning to bring this up at the IPT meeting tomorrow to see if anyone else has more context.
For category #1 (e.g. XXX-XX-XXXX), on our end, we can simply strip out any dashes in file numbers we get from BGS. (See https://github.com/department-of-veterans-affairs/vets-api/pull/12530). However, like the bad SSN problem described previously, there is nothing we can do about categories #2 and #3.

**Comment by data-doge** [comment]
I drafted https://github.com/department-of-veterans-affairs/vets-api/pull/12530 which will resolve some of the file number mismatches, and improves our logging around this issue for our convenience, and to help us report on mismatches to MPI and BGS, so that they can follow up. Once that PR is merged, I think we can close this ticket.

**Comment by saderagsdale** [comment]
@data-doge Can you estimate this since it includes code work?

**Comment by data-doge** [comment]
https://github.com/department-of-veterans-affairs/vets-api/pull/12530 is working as expected. We can close this ticket in favor of https://github.com/department-of-veterans-affairs/va.gov-team/issues/57798
